Enable matmul scheduler in segmenter

drzejan2 opened the pull request:
PR to track progress of enabling matmul scheduler in segmenter.

A proper description will be added when PR will be closer state in which it can be reviewed.

drzejan2 committed:
Segmenter - support for matmul scheduler

- connect matmul scheduler in segmenter with implementation of
  matmul scheduling,
- update matmul params to store key items needed for matmul scheduling
  (based on prototype param structure),
- add a modified copy of schedule matmul that has API used by
  segmenter,
- update implementation of tests cases to use updated infrastructure
  for scheduling matmuls (previously it was a prototype),
- move compile/runtime checks to separete source file,

drzejan2 committed:
Test case for checking matmul scheduler in segmenter

csarofeen commented on third_party/nvfuser/csrc/scheduler/registry.cpp:
```diff
@@ -2038,10 +2044,272 @@ class PersistentKernelScheduler : public SchedulerEntry {
+        if (mmaExprInputs.size() != 2) {
+          scheduler_debug_utils::canScheduleRejectReason(
+              ScheduleHeuristic::Matmul,
+              "One of mma op doesn't have 2 inputs, got: " +
```
Can the mma op be defined without two input tensor views?

drzejan2 replied:
That's a good question. I can think of situation where the same TV is used for both inputs but currently `scheduleMartmul` does not support this configuration. My idea for now is to have a skeleton that matches what is expected as of today in matmul tools and extend it when more details are available.

csarofeen replied:
I don't think these checks are needed since mma op requires explicitly two inputs one output all TensorViews: https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/ir_nodes.cpp#L1324-L1336

drzejan2 replied:
Thanks for pointer to the definition of mma ir node, I will remove this check from here as it is redundant at this point.

csarofeen commented on third_party/nvfuser/csrc/scheduler/registry.cpp:
```diff
@@ -2038,10 +2044,272 @@ class PersistentKernelScheduler : public SchedulerEntry {
+        }
+
+        // MMA input TVs must be in local memory
+        // NOTE: shouldn't buffers be in shared memory?
```
I think what you should be checking here is if the inputs are an input of the fusion. Memory type can be changed for anything that's not input/output to any memory space. Fusion inputs/outputs must be in global memory.

drzejan2 replied:
I wasn't sure about how deep the verification should go, I though about 'basic' set of checks:

1/ check if there are any mma ops (I assume this is required for matmul)
2/ access mma op input TVs and check IterDomains to see which are broadcast - to match data layout

The follow up checks would be:
3/ if mma op inputs are based on fusion inputs (but this might be redundant)
4/ if mma op output is fusion output / is used to produce fusion output in case if there is epilogue operation such as bias

naoyam replied:
Why are they be in Local memory? Aren't we assuming they are in global memory?

csarofeen replied:
yeah fusion->exprs() will filter out anything not connected to outputs/inputs so 3 shouldn't be necessary, but might want to check the explicit pattern of input->cast->broadcast->mma, but that's only fully necessary when we allow mma translation in the E2E stack. For now I think these checks are generally good.

But yes, generally you've got the full idea. This check is to validate/limit exactly what the scheduler gets. So when we go to deploy it E2E it should reject anything the scheduler can't handle (in combination with the runtime check).

csarofeen reviewed (commented):
This looks like good progress, just adding a couple minor comments for now.

Self note: Since we don't do any translation to an MMA op we don't have to hide scheduling with an env variable. Once we start translating we should hide the translation under an env variable with: https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/utils.h#L110-L119

drzejan2 commented:
@zasdfgbnm, @naoyam 
Can I ask you to check the initial version of `MatmulScheduler::canScheduleCompileTime` in `registry.cpp`? Does it go in a reasonable way or the checks are too heavy and should be simplified?

I plan to drop checks for memory type for mma op, as @csarofeen suggested.

csarofeen commented on third_party/nvfuser/csrc/scheduler/registry.cpp:
```diff
@@ -2038,10 +2044,272 @@ class PersistentKernelScheduler : public SchedulerEntry {
+          mmaExprs.push_back(expr);
+        }
+      }
+    }
```
slightly cleaner syntax would be:
`auto mmaExprs = ir_utils::filterByType<MmaOp>(fusion->exprs()).vector();`

drzejan2 replied:
Applied in the latest revision, marking as resolved.

csarofeen commented on third_party/nvfuser/csrc/scheduler/registry.cpp:
```diff
@@ -2038,10 +2044,272 @@ class PersistentKernelScheduler : public SchedulerEntry {
+
+    // The plan:
+    // 1. check if MmaOp is defined
+    // 2. check if TV bound to MmaOp is in outputs
```
This check is unnecessary by definition of mma op:
https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/ir_nodes.cpp#L1324-L1336

drzejan2 replied:
I fixed that in the updated version, which I will push shortly.
Thanks!

csarofeen commented on third_party/nvfuser/csrc/scheduler/registry.cpp:
```diff
@@ -2038,10 +2044,272 @@ class PersistentKernelScheduler : public SchedulerEntry {
+    // The plan:
+    // 1. check if MmaOp is defined
+    // 2. check if TV bound to MmaOp is in outputs
+    // 3. check if inputs for output are in local memory
```
This check should just be that input and ouputs are not fusion inputs/outputs

drzejan2 replied:
I address this in the following [comment](https://github.com/csarofeen/pytorch/pull/2461#discussion_r1120159729), I'm marking this comment as resolved to merge these two conversations.

naoyam commented on third_party/nvfuser/csrc/scheduler/registry.cpp:
```diff
@@ -2038,10 +2044,272 @@ class PersistentKernelScheduler : public SchedulerEntry {
+        // NOTE: shouldn't buffers be in shared memory?
+        if (inA->getMemoryType() != MemoryType::Local ||
+            inB->getMemoryType() != MemoryType::Local) {
+          const auto memTypeToString =
+              [](const MemoryType memoryType) -> std::string {
+            switch (memoryType) {
+              case MemoryType::Global:
+                return "GLOBAL";
+              case MemoryType::Local:
+                return "LOCAL";
+              case MemoryType::Shared:
+                return "SHARED";
+            }
```
nit: `MemoryType` can just be printed as there's `operator<<` overload for `MemoryType`.

drzejan2 replied:
Thanks!
I will fix this in the following updates of this MR.

naoyam commented:
> @zasdfgbnm, @naoyam Can I ask you to check the initial version of `MatmulScheduler::canScheduleCompileTime` in `registry.cpp`? Does it go in a reasonable way or the checks are too heavy and should be simplified?
> 
> I plan to drop checks for memory type for mma op, as @csarofeen suggested.

Just added a couple of comments. Looks mostly good to me.

csarofeen commented on third_party/nvfuser/csrc/scheduler/registry.cpp:
```diff
@@ -2038,10 +2044,272 @@ class PersistentKernelScheduler : public SchedulerEntry {
+            outputs.cend()) {
+          mmaOutputs.push_back(expr);
+        }
+      }
```
You can just check ->isFusionOutput/Input directly on the tensor views.

drzejan2 replied:
With how we do matmul at the moment ([link](https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/test/test_utils.cpp#L253)) this check will not help us. The MmaOp direct inputs are a result of already processed fusion inputs which are marked to be in local memory, to answer @naoyam question why this is not in global memory.

I'm thinking about tracking down if MmaOP inputs are a result of some operations on fusion input - this might be helpful in future if we would like to add support for conversion (e.g. fp32 -> tf32).
It is similar situation for MmaOp output, with prepared tests it is indeed an output of fusion but if we were to support full alpha:

```
C = alpha*AxB
```
then the result of MmaOP will be a partial data, used to multiply by alpha to produce the final result.

To have it done I will need to build a 'map' with relationships (producer-consumer) between Vals with MmaOp in the center of this map (I boldly assume there is a single MmaOp in the fusion).

My major concern is that this map will have a fixed 'routes' for predefined set of expressions. On the other hand, defining checks for specific patterns/routes in the Vals' dependencies may be useful for adding extended functionality (mentioned alpha or beta with additional input tensor, later also epilogues).

An example of such a route could be
- for vanilla matmul (things to be checked):
-- MmaOP output is fusion output,
-- MmaOp inputs are a result of adding broadcast to the input TVs
-- fusion has two inputs,
-- fusion has one output,
- for alpha matmul:
-- fusion has two inputs,
-- fusion has one output,
-- fusion output is a consumer of product of MmaOp output (matmul result) and 3rd fusion input (alpha)
-- MmaOp inputs are a result of adding broadcast to input TVs,

Does it make sense? There is a risk that such a pattern matching might be expensive, even for compile time processing.

csarofeen replied:
I don't think this should be expensive. I guess maybe the checks though I was thinking (maybe not immediately) would be:
-- Broadcast pattern of MmaOp (I think we can just check where the broadcast positions are relative to the two inputs of the MMA Op)
-- All operations on inputs to MmaOp are Unary ops
-- All operations on outputs of MmaOp are Unary ops

I think unary ops should be quite straightforward to process in the Mma scheduler, but maybe that's not the case?

csarofeen reviewed (commented):
Generally looking good. I don't think these checks are too heavy weight, and the compile time check is intended for the heavy weight checks. Added some comments, nothing major.

drzejan2 commented on third_party/nvfuser/csrc/scheduler/registry.cpp:
```diff
@@ -14,6 +14,12 @@
 #include <limits>
+#include <utility>
+#include "c10/util/Exception.h"
+// #include "ir_interface_nodes.h"
```
TODO: to be removed.

drzejan2 committed:
Segmenter - support for matmul scheduler

- enable matmul scheduler in segmenter, simple test passes,
- add initial heuristicis for matmul scheduler in segmenter

csarofeen commented on third_party/nvfuser/csrc/scheduler/matmul.cpp:
```diff
@@ -320,4 +329,305 @@ void scheduleMatmul(
+  // Different paths because Volta swizzle needs to
+  //  involve the broadcast dimensions that are concretized
+  //  at mma, while Ampere ones should be done before
+  //  the broadcast op to be able to use cp.async.
```
What's the limitation here? Is this because of our implementation or something more fundamental?

drzejan2 replied:
Unfortunately, I have dived into documentation of both archs to tell what was the reason for this split. Shiming left only this comment here.

csarofeen replied:
Not a big deal, don't dive too deep right now ;-)

csarofeen reviewed (commented):
This looks like it's coming along well, just pushing a comment.

drzejan2 committed:
Matmul scheduler - apply changes from #2488

- apply improvement in matmul instruction scheduling with loop rotation

drzejan2 commented on third_party/nvfuser/csrc/scheduler/matmul_heuristic.h:
```diff
@@ -0,0 +1,67 @@
+      return false;
+    }
+    // TODO: to be changed when there are some parameters
+    return false;
```
@drzejan2  - add missing implementation.

drzejan2 commented on third_party/nvfuser/csrc/scheduler/registry.cpp:
```diff
@@ -2306,6 +2385,10 @@ void HeuristicSummary::validate() const {
       break;
     }
+    case ScheduleHeuristic::Matmul: {
+      // TODO: add proper set of checks
```
@drzejan2 - remove comment and create a ticket for it.

drzejan2 replied:
Fixed in later change.

drzejan2 commented on third_party/nvfuser/csrc/scheduler/matmul_checks.cpp:
```diff
@@ -0,0 +1,478 @@
+        instruction_tile.k * k_ratio};
+  }
+
+  // TODO: support for Volta and Turing
```
@drzejan2 - remove comment and add ticket for tracking

drzejan2 replied:
Created a ticket to track this change.

drzejan2 committed:
Segmenter - support for matmul scheduler

- add implementation of helper functions for matmul heuristics,
- add dedicated debug logger,
- add test for checking matmul schedule integration with segmenter,

csarofeen commented on third_party/nvfuser/csrc/scheduler/matmul.cpp:
```diff
@@ -325,4 +334,310 @@ void scheduleMatmul(
+      outputs[0]->isA<TensorView>(),
+      "fusion's output is not an instance of TensorView class");
+
+  TensorView* a = dynamic_cast<TensorView*>(inputs[0]);
```
Check out https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/utils.h#L153-L235

You can just do `inputs[0]->as<TensorView>()` with our IR nodes

drzejan2 replied:
Thanks for pointing this - I need to get more used to these tools so.
I fixed this in PR that I created in `NVIDIA/Fuser` project ([link](https://github.com/NVIDIA/Fuser/pull/23))

csarofeen commented on third_party/nvfuser/csrc/scheduler/matmul.cpp:
```diff
@@ -325,4 +334,310 @@ void scheduleMatmul(
+
+  // TODO:
+  // Beyond this point, mma_builder really just becomes a populated
+  //  list of parameters to describes the mma swizzles that should
```
nit: describe

drzejan2 replied:
Fixed in the updated sources, that are present in PR in new project ([link](https://github.com/NVIDIA/Fuser/pull/23))

csarofeen commented on third_party/nvfuser/csrc/scheduler/matmul.cpp:
```diff
@@ -325,4 +334,310 @@ void scheduleMatmul(
+  // Get the input to the mma op.
+  auto mma = dynamic_cast<MmaOp*>(cc->definition());
+  TORCH_INTERNAL_ASSERT(mma != nullptr);
+  auto ab = mma->inA()->as<TensorView>();
```
note here. so `cc->definition()->as<MmaOp>()` for the above dynamic cast too

drzejan2 replied:
Fixed in a following comment - it will be visible in new PR ([link](https://github.com/NVIDIA/Fuser/pull/23))

csarofeen commented on third_party/nvfuser/csrc/scheduler/matmul_utils.cpp:
```diff
@@ -0,0 +1,596 @@
+
+  // #1
+  auto path_to_mma_A =
+      DependencyCheck::getAllDependencyChains(fusion_inputs[0], mma_inputs[0]);
```
maybe it'd be better to grab all paths from all fusion inputs to the mma_inputs[0]. Then you could check if the mma_inputs[0] or mma_inputs[1] is in that set. Similar for output checks. You could also make sure there's no intersection between the sets i.e. no intersection of the sets {dependency check from fusion inputs to the first mma input, and the fusion inputs to the second mma input}. Similar thing could be checked for epilogue, this might help you isolate the different portions of the fusion.

drzejan2 replied:
@csarofeen I will check your idea with goal of creating a helper tool that can be reused in future for other matmul patterns. There is one thing that I will need to check - how dependencies between inputs look if fusion were to represent `C = alpha * A x B` set of expressions.

I expect that in scenario there will be intersection in paths to mma 1st input between fusion 1st input (alpha) and its 2nd input (A)

csarofeen replied:
Yeah, only recommendation I have is thoroughly read through iter_visit.cpp and scheduler/registry.cpp . Those have a lot of very specific graph checking logic and should give you a feel for some different ways to think/look through the graph.

For `C = alpha * A x B` I would think you'd just look through the history of A and check that the only branches are from scalar expressions otherwise you're on the path to the fusion input leading to A.

drzejan2 replied:
I will revisit files you mentioned as I now understand that my approach is a little bit too strict / not abstract enough. I realized this when I though how to handle alpha. I focused too much on a single pattern even though this functionality could be expressed in the following ways:
```
C = (alpha * A) x B
C = A x (alpha * B)
C = alpha * (A x B)
```
so the current compile-time check needs redoing.

Thanks for pointing this out!

csarofeen reviewed (commented):
Didn't get through the whole PR but pushing a few comments before our sync today.

drzejan2 commented:
I opened PR with the same set of changes in NVIDIA/Fuser project - [here link to new it](https://github.com/NVIDIA/Fuser/pull/23).
When I address all comments in this PR I will close it.

naoyam commented on third_party/nvfuser/csrc/mma_type.cpp:
```diff
@@ -112,12 +113,15 @@ bool isVolta(MmaOptions::MacroType macro) {
 
 bool isTuring(MmaOptions::MacroType macro) {
-  return macro == MmaOptions::MacroType::Turing_16_8_16 ||
+  return false || //
```
Why `false ||`?

drzejan2 replied:
This is indeed unneeded, I added it to control code formatting. I will remove it (here and below).

Thanks!

naoyam commented on third_party/nvfuser/test/test_gpu_tensorcore.cpp:
```diff
@@ -3053,6 +3029,84 @@ TEST_F(NVFuserTest, FusionAmpereMatmulLargeLoadLargeK_CUDA) {
+        !executor_cache.getMostRecentKernelRuntime()->isSegmented(),
+        "segmentation did happen");
+
+    // NOTE: checking with lower expectations for relative/absolute error
```
Does the validation fail if compared against a reference result computed using double-precision inputs?

drzejan2 replied:
The reference is a result of using fp16 inputs.
I checked the already existing set of matmul schedule tests and in all of them the value of acceptable error has been set explicitly.

naoyam commented on third_party/nvfuser/csrc/scheduler/matmul_utils.cpp:
```diff
@@ -0,0 +1,596 @@
+using TensorShape = std::vector<int64_t>;
+using ProblemShape = TensorShape;
+
+//! A constant with position of M value (a number of columns in A tensor for TT
+//! layout) in problem in ProblemShape type.
+constexpr size_t M_POS = 0;
+//! A constant with position of N value (a number of rows in B tensor for TT
+//! layout) in problem in ProblemShape type.
+constexpr size_t N_POS = 1;
+//! A constant with position of K value (a number of rows in A tensor for TT
+//! layout) in problem in ProblemShape type.
+constexpr size_t K_POS = 2;
+//! A constant with expected number of dimensions in ProblemShape type.
+constexpr size_t PROBLEM_DIMS = 3;
```
nit: Are these declarations used outside of this file? Can they be moved to the anonymous namespace below?

drzejan2 replied:
A valid point - I will double check if all of them are still in use (there were some major changes in design) and will move what's left to anonymous namespace.

Thanks!

naoyam commented on third_party/nvfuser/csrc/scheduler/matmul_utils.cpp:
```diff
@@ -0,0 +1,596 @@
+}
+
+//! A helper for deciding what kernel indexing mode use (int32_t or int64_t).
+PrimDataType getIndexType(const ProblemShape& problem_shape) {
```
Is this necessary because `collectIndexMode` does not check fusion outputs and intermediates?

drzejan2 replied:
I added this when indexing mode has become required (at least set) in heuristics params. It seems that `collectIndexMode` was not triggered anywhere in the execution flow and I started getting asserts in `kernel_cache.cpp:368`, in `FusionKernelRuntime::runKernelWithInput` function.

If the check on input params should be done in a different place, then I'm ok to remove it from here and use the expected execution flow.

naoyam replied:
`collectIndexMode` is called from `KernelArgumentHolder::createKernelArgumentHolder`, which is called from `FusionExecutorCache::prepareInputs`, and which is called from `FusionExecutorCache::runFusionWithInputs`. As long as `runFusionWithInputs` is used, `collectIndexMode` should be called.

I don't think `FusioonKernelRuntime::runKernelWithInput` should be used. It was added to support async compilations, but I'm not sure if it's still maintained. We should explicitly mark it as *EXPERIMENTAL*. Please use `FusionExecutorCache::runFusionWithInputs`.

naoyam commented on third_party/nvfuser/csrc/scheduler/matmul.cpp:
```diff
@@ -325,4 +334,310 @@ void scheduleMatmul(
+  const auto& inputs = fusion->inputs();
+  const auto& outputs = fusion->outputs();
+
+  TORCH_CHECK(
```
nit: Please use `TORCH_INTERNAL_ASSERT` instead of `TORCH_CHECK` for anything that's not directly called by users.

drzejan2 replied:
I updated sources to use `TORCH_INTERNAL_ASSERT` in changed sources. I will follow this pattern in future changes, thanks!

naoyam commented on third_party/nvfuser/csrc/scheduler/registry.cpp:
```diff
@@ -2034,10 +2035,67 @@ class PersistentKernelScheduler : public SchedulerEntry {
+      return false;
+    }
+
+    const auto msg = getMatmulCompileTimeRejectReason(fusion);
```
Like this abstraction 👍

naoyam commented on third_party/nvfuser/csrc/scheduler/registry.cpp:
```diff
@@ -2034,10 +2035,67 @@ class PersistentKernelScheduler : public SchedulerEntry {
+        getMatmulRunTimeRejectReason(fusion, data_cache, runtime_info);
+    if (!reason.empty()) {
+      scheduler_debug_utils::canScheduleRejectReason(
+          ScheduleHeuristic::Transpose, reason);
```
```suggestion
          ScheduleHeuristic::Matmul, reason);
```

drzejan2 replied:
Fixed in the PR in new project. Thanks for catching this.

naoyam commented on third_party/nvfuser/test/test_gpu_tensorcore.cpp:
```diff
@@ -3053,6 +3029,84 @@ TEST_F(NVFuserTest, FusionAmpereMatmulLargeLoadLargeK_CUDA) {
+
+  auto outputs = executor_cache.runFusionWithInputs({t0, t1});
+
+  TORCH_CHECK(
```
Also assert that the matmul scheduler is indeed used.

drzejan2 replied:
Good point, I assumed it works as running this test without matmul schedule ends with assertion triggered by pointwise scheduler.

I added an utility function for detecting scheduler use in fusion and updated prepared test. Changes are available in PR made in new project ([link](https://github.com/NVIDIA/Fuser/pull/23))

naoyam commented:
I don't see any red flag. I just left general comments. I'm OK to get it merged. Is it still a draft?

drzejan2 commented:
@naoyam I kept this PR in draft state for the time being, when I ported changes to project's new location. 

I applied changes from comments in newly opened PR in project's new localization ([link](https://github.com/NVIDIA/Fuser/pull/23)).

To avoid confusion, I'm closing this PR.
